feat: :sparkles: `explain()` errors

joelostblom opened the pull request:
Creates a more human readable format to explain error messages

Closes #11, closes #58, closes #14

Needs an in-depth review.

## Checklist

- [x] Ran `just run-all`

joelostblom committed:
feat: Create explain function to improve error formatting

Hopefully makes errors easier to read

joelostblom committed:
feat: Print how many issues were found

joelostblom commented on src/check_datapackage/issue.py:
```diff
@@ -31,3 +32,4 @@ class Issue:
     jsonpath: str
     type: str
     message: str
+    instance: Any
```
We could also enumerate the possible types here, but I'm not 100% on what they should be. From running

```sh
rg "\"type\": \"" ./src/check_datapackage/schemas/data-package-2-0.json |tr -d " " | tr -d ',' | sort | uniq
```

I see:

```
"type":"array"
"type":"boolean"
"type":"integer"
"type":"number"
"type":"object"
"type":"string"
```

It seems like `array` could be `List`, but what is `object`? Maybe `Dict`?

lwjohnst86 commented on src/check_datapackage/check.py:
```diff
@@ -52,15 +52,34 @@ def __init__(
+        "|\n"
+        f"| {property_name}: {issue.instance!r}\n"
+        f"| {' ' * len(property_name)}  {'^' * len(str(issue.instance))}\n"
+        f"{issue.message}\n"
```
I think we might need to have some custom messages, as some of the messages are not helpful (e.g. the regex ones).

But that can be another PR :relaxed:

lwjohnst86 reviewed (changes requested):
Super cool :tada: Some small comments so far :pen:

lwjohnst86 committed:
Merge branch 'main' into feat/explain-errors

joelostblom committed:
refactor: Move message creation into separate function

Moves all logic out of init()
and instead into two separate functions:
one for iterating over the errors
and adding a general message,
and one for formatting the message
for each issue.

lwjohnst86 commented on src/check_datapackage/check.py:
```diff
@@ -48,17 +48,40 @@ def __init__(
+    """
+    issue_messages: list[str] = _map(
+        issues,
+        _create_issue_message,
```
```suggestion
        _create_explain_message,
```

lwjohnst86 reviewed (changes requested):
Nice! Very minor comment about naming :relaxed:

joelostblom committed:
refactor: Rename for clarity

joelostblom committed:
fix: Shorten line

joelostblom committed:
feat: Set default value for issue

To avoid having to specify it everywhere where it is not used.

joelostblom commented on src/check_datapackage/extensions.py:
```diff
@@ -162,6 +163,7 @@ class RequiredCheck(BaseModel, frozen=True):
     jsonpath: JsonPath
     message: str
     _targets: list[TargetJsonPath] = PrivateAttr()
+    instance: Any = None
```
@martonvago @lwjohnst86 I set the default to `None` so that we don't have to pass `error.instance` around everywhere that we create an Issue (since it is only used for the `explain` message right now), but if you think that approach is better, I can change to that.

lwjohnst86 replied:
Hmm, thinking about this more, if `instance` is part of the `explain` message, we should actually require that it be present, no? Rather than default to `None`?

joelostblom replied:
Ok, tomorrow I can look into the other places where it is used and see if it makes sense to always include a message there. For example for the required and [custom check classes](https://github.com/seedcase-project/check-datapackage/blob/9410ee4881207ae124d86aeacefe4a4e4154a550/src/check_datapackage/extensions.py#L92-L94) (maybe you know already if that makes sense?).

lwjohnst86 replied:
Needs a docstring for this attribute.

joelostblom commented on src/check_datapackage/check.py:
```diff
@@ -579,6 +602,7 @@ def _create_issue(error: SchemaError) -> Issue:
         message=error.message,
         jsonpath=error.jsonpath,
         type=error.type,
+        instance=error.instance,
```
After some testing, it seems like any properties dictionary/json that misses the top level `resources` attribute triggers this error... :thinking:

lwjohnst86 replied:
Hmmm, maybe for those cases, don't include the instance and leave it `''` blank? :thinking:

martonvago replied:
Yeah, exactly what I was thinking too! And yeah, I wasn't sure how you wanted to handle these bulky `instance`s in the explain output. If you don't want to show them at all, then weeding them out earlier is more practical. But, conceptually at least, removing `instance` from the hash and handling `instance`s we don't want to display as part of the display logic feels cleaner to me. It just feels a bit off to be exposing `instance` to the user and not have it reliably contain the instance. 🤔

joelostblom committed:
fix: Only define `instance` when it is a hashable type

For certain error types,
a dict or list is set as the instance,
which means logic elsewhere
that calls `set()` on `Issue` fails.

lwjohnst86 commented on src/check_datapackage/check.py:
```diff
@@ -48,17 +48,40 @@ def __init__(
+    """Create an informative explanation of what went wrong in each issue."""
+    property_name = issue.jsonpath.split(".")[-1]
+    return (  # noqa: F401
+        f"At package.{issue.jsonpath[2:]}:\n"  # indexing to remove '$.'
```
```suggestion
        f"At `package.{issue.jsonpath[2:]}`:\n"  # indexing to remove '$.'
```

I think backticks or some other visual separator would be nice here.

joelostblom replied:
Personally, I think it can look a bit cluttered with raw markup being printed as output. Could we punt this to a PR that deals with the styles of these messages and considers using rich? E.g. just using the `print` from rich would give us some additional formatting:

<img width="450" height="274" alt="image" src="https://github.com/user-attachments/assets/8c51d308-836a-46dd-b467-30173b001529" />


Rich also does pretty printing of json objects and creates stylized tracebacks, such as this example:

<img width="926" height="369" alt="image" src="https://github.com/user-attachments/assets/d59000a3-4d5a-488a-90d0-0efc3f2517fb" />

lwjohnst86 commented on src/check_datapackage/check.py:
```diff
@@ -48,17 +48,40 @@ def __init__(
+    return (  # noqa: F401
+        f"At package.{issue.jsonpath[2:]}:\n"  # indexing to remove '$.'
+        "|\n"
+        f"| {property_name}: {issue.instance!r}\n"
```
Should we wrap `property_name` in `""` to match JSON style or match Python style? :thinking:

joelostblom committed:
fix: Only add the error instance if it is hashable

Since we perform deduplication of Issues later,
we need to either only allow hashable items
or write custome comparison logic
to compare unhashable ones.
Technically,
writing custom comparison logic
that ignores unhashable values
might be the most theoretically appropriate
(since it is redundant to compare `instance`
when also comparing `jsonpath`),
but it could look convoluted
and have worse performance.
This is a simpler solution
that is easy to understand
and that does not involve any significant performance penalty
as the extra comparison is of small items.
Trying this out
as a first approach
and changing if we notice that there are occasions
where we would value having access to an unhashable instance.
For more discussion, see:
https://github.com/seedcase-project/check-datapackage/pull/208#discussion_r2555916842

lwjohnst86 committed:
Merge branch 'main' into feat/explain-errors

joelostblom committed:
feat: Expose explain at top level

joelostblom committed:
docs: Add explain example

joelostblom committed:
docs: Show sections on `explain`

lwjohnst86 reviewed (changes requested):
Look super nice! Just some small comments about docstrings.

joelostblom committed:
fix: Remove unnecessary `instance` initializations

joelostblom committed:
docs: Add docstring

joelostblom committed:
fix: Account for quotes when printing string instances

lwjohnst86 committed:
Merge branch 'main' into feat/explain-errors

joelostblom committed:
feat: Exclude Issue instance from comparison and hash

But always include it in the Issue object
instead of sometimes excluding it
based on whether it is hashable or not.
See further discussion in
https://github.com/seedcase-project/check-datapackage/pull/208#discussion_r2555916842

martonvago commented on src/check_datapackage/check.py:
```diff
@@ -48,17 +48,55 @@ def __init__(
+    # Two extra carets are added for string instances to account for quotes
+    number_of_carets = len(str(issue.instance)) + 2 * isinstance(issue.instance, str)
+    return (  # noqa: F401
+        f"At package.{issue.jsonpath[2:]}:\n"  # indexing to remove '$.'
```
Can also do `issue.jsonpath.removeprefix("$.")` and then maybe the comment is not necessary

martonvago commented on src/check_datapackage/check.py:
Here, we rewrite `jsonpath` so that for `type="required"` errors it ends in the missing field's name. I think that (regardless of having solved the hash problem) this should be matched by rewriting `instance` for required errors in the same way. Otherwise, we will end up with output like:
```
# check({"resources": [{"title": "Title"}]}, error=True)

At package.resources[0].name:
|
| name: {'title': 'Title'}
|       ^^^^^^^^^^^^^^^^^^
'name' is a required property
```

It would be nice if this could happen here and not in `_create_issue`, but I don't know off the top of my head if that would be problematic for any logic that relies on `instance` on `SchemaError`s.

joelostblom replied:
Interesting, do you think this is something that can be solved in a separate PR? It's not immediately clear how to differentiate the two errors I get from running your example:

<img width="843" height="507" alt="Image" src="https://github.com/user-attachments/assets/47bbe565-18ab-4e11-8d2f-c7a9ff5d7955" />

The first one seems correct whereas the second one has the issue you highlighted.

martonvago reviewed (dismissed):
Really nice ⭐!!

I added some comments, feel free to ignore what you don't care about ☺️

joelostblom committed:
fix: Fix typos

Co-authored-by: martonvago <57952344+martonvago@users.noreply.github.com>

joelostblom committed:
fix: Skip unnecessary variable assignment

joelostblom committed:
fix: Correct grammatical number

joelostblom committed:
feat: Account for `[]` passed as the object

joelostblom committed:
fix: Don't include extra carets since we no longer have the quotes printed

joelostblom reviewed (commented):
Thanks @martonvago ! I went through and tried to address your comments.

lwjohnst86 reviewed (approved):
Very nice :tada:
